[Chew Hong Jin] iP

sarrrdin opened the pull request:
#DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind from having to remember things you need to do. DukePro is:

- text-based
- easy to learn
- ~~FAST~~ **_ULTRA FAST_** to use

All you need to do is,

1. download it from [here.](https://github.com/sarrrdin/ip)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😄 

And it is **FREE OF CHARGE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here is the ```main``` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

sarrrdin committed:
Completed Level - 1

sarrrdin committed:
Added one new condition, when user inputs "list" ; change method when user inputs a task

sarrrdin committed:
Created new Class to store Task

sarrrdin committed:
Created new Class TaskList to store list of tasks

sarrrdin committed:
added new condition when user inputs list and change method when user input a task

sarrrdin committed:
Added Condition when user input mark or unmark

sarrrdin committed:
changed method getTaskList to check on whether task is done or not done, also added method to mark and unmark

sarrrdin committed:
Changed class documentation, added methods to mark and unmark specific tasks in list based on index

sarrrdin committed:
Change if else to switch instead for elegance, also set initial and bye message as final.

sarrrdin committed:
Add additional variable - type of task.

sarrrdin committed:
Changed implementation such that Task class handles all String output/message based on task type.

sarrrdin committed:
Changed arguments for Task class constructor based on change in task constructor.

sarrrdin committed:
change in message output.

sarrrdin committed:
added a few exception handling,,, will improvise if have time by creating new exception class to handle errors. exception handled include length of list; invalid argument; not sure if description of deadline and event is required.

sarrrdin committed:
Changed Mark and unmark method, realise opposite commadn put lol

sarrrdin committed:
Edited the mark part where it displays which task is marked/unmarked

sarrrdin committed:
added deletion method, at TaskList class too to handle removal of task

sarrrdin committed:
Added new class DukeException to handle Exceptions thrown for illegal argument/non-existent command, index out of range, as well as empty description

sarrrdin committed:
Added some JavaDocs, improvised Exception handling

sarrrdin committed:
Added test cases

sarrrdin committed:
Created Enum Class use Enums

sarrrdin committed:
Add Save File option

sarrrdin committed:
Commit Save File Function

sarrrdin committed:
import date function

sarrrdin committed:
import datetimeformatter, will edit oop later

sarrrdin committed:
Merge branch 'branch-Level-7'

sarrrdin committed:
Changed implementation by for every case, a new instance of the object is created, new class that extends Tasks - Deadline, Event, Todo

sarrrdin committed:
Merge branch 'branch-Level-8'

sarrrdin committed:
Added Command Classes for greater use of OOP

sarrrdin committed:
inform git to ignore class giles (ahh i think im abit late)

sarrrdin committed:
Created Command PAckage, new classes: UI to handle user interface, parser to handle input, changed implementation of main class at Duke.java

sarrrdin committed:
divide classes into package, create new package task to store all task classes: Deadline, Event, Todo, Task; made Task abstract class

sarrrdin committed:
in the midst of figuring out how gradle works...will get back to this later

sarrrdin committed:
added Gradle into current project

sarrrdin committed:
Added JUnit Test Cases, also did abit of java doc while figuring out how this JUnit thing works.

sarrrdin committed:
edited gitignore file to ignore .jar file so wont commit

sarrrdin committed:
tidy up code, remove all unecessary Imports

sarrrdin committed:
Added Javadocs for all Command Classes (did not add last time)

sarrrdin committed:
Change method name and switch format to follow java Coding Standard

sarrrdin committed:
Add find function for Duke chatbot

sarrrdin committed:
Merge branch 'branch-A-CodingStandard'

sarrrdin committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java

sarrrdin committed:
Fix bug for compilation error in FindCommand after changing printMessage

sarrrdin committed:
Added Gradle Function to checkstyle, will code style later.

sarrrdin committed:
Add varargs for find function, where it can take in a list of string.

nmtuan2001 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,54 @@
+ */
+package duke;
+
+import duke.command.*;
```
Maybe you can import each class you need instead of using *?

sarrrdin replied:
oh yes, thanks for pointing that out :)

nmtuan2001 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,126 @@
+     */
+    public static Command parse(String fullCommand) throws DukeException {
+
+        String[] com = fullCommand.split(" ", 2);
```
Should this variable name be plural?

sarrrdin replied:
updated

nmtuan2001 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,100 @@
+     */
+    public ArrayList<Task> loadFromFile() {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(" d/MM/uuuu");
+        ArrayList<Task> memo = new ArrayList<>();
```
Maybe this attribute should be plural too?

sarrrdin replied:
oh yes, thanks for pointing that out :)

nmtuan2001 reviewed (commented):
There aren't a lot of things for me to comment tbh. Looks great :thumbsup: :thumbsup:

sarrrdin committed:
Trying out GUI for Duke.

sarrrdin committed:
Change some minor formatting

sarrrdin committed:
Fix final bug on app.

sarrrdin committed:
Edit based on comments given by peer

Sampy147 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,99 @@
+/**
+ * public class Ui to handle text shown to users when command is typed in to Duke.
+ */
+public class Ui {
```
Good idea to add a UI class, definitely helps in handling functions that prints out the response

Sampy147 commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,38 @@
+     * @param storage
+     */
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
```
Good idea to have subclasses of Command have an execute method. Ensures consistency across all subclasses and improves readability

Sampy147 commented on src/test/java/duke/task/EventTest.java:
```diff
@@ -0,0 +1,22 @@
+
+public class EventTest {
+    @Test
+    public void test1() {
```
Perhaps a good idea to give a brief idea of the test in the method name of the test rather than "test1()" so that readers can understand the general idea of what the test does

Sampy147 reviewed (commented):
The way OOP was applied was really good, and the way the classes were structured was great! Keep up the good work

Qiaoran-M commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,123 @@
+     * @return Arraylist<String> containing tasks in String form.
+     */
+    public ArrayList<String> saveList() {
+        ArrayList<String> result = new ArrayList<>();
```
result variable can be named plural.

Qiaoran-M commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * public abstract class Command that checks command.
+ */
+public abstract class Command {
```
Good OOP design for Command class!

Qiaoran-M commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+package duke.task;
+
+import java.time.LocalDate;
+import java.time.format.*;
```
You may consider list the imported classes explicitly.

Qiaoran-M commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,38 @@
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
+        storage.saveToFile(tasks.saveList());
+        return ui.printMark(tasks.get(index));
```
Have you considered throwing index out of bound DukeException for mark and delete command? (pls ignore this comment if you have written it somewhere and I missed it :-))

Qiaoran-M reviewed (commented):
Overall, this pretty much LGTM and the OOP structure is well-designed!

sarrrdin committed:
DialogBox: after every command, dialogbox is unable to show full
message

DialogBox: DialogBox should display full command and output

This helps user to fully comprehend what is going on.

GUI cannot show full message due to height and width restrictions,
hence updated DialogBox to achieve so

sarrrdin committed:
UI: the lines are present even in GUI

The line may be distraction for users when they read the commands.

Having the line absent may be more user friendly because user can
clearly read the output of Duke and the progress of their tasks.

Let's remove the lines included and tidy up the code.

sarrrdin committed:
#Use Assertions in TaskList and Commands
The Commands given by user, handled with methods in TaskList deals
with requests from user. They may fail if parameters are invalid.

Adding Assertions to check if input index are valid and check for
assumptions about the algorithm helps runtime to verify if input
are valid. This helps programmer detect mistake.

Let's use Assertions to assist in identifying bugs in code.

sarrrdin committed:
#Improve Code Quality

Many public classes and method not only do not have the relevant
javaDocs or comments. They violate the correct Java Coding Standard
too.

Ensuring that these coding styles are checked and rectified helps
future developers comprehend code better as it improves readability.

The addition and rectification of code makes it easier for future
improvements for understanding the logic of algorithm.

Let's,
*rectify relevant violations
*structure code well and add relevant javaDocs

sarrrdin committed:
Use streams in TaskList specifically findTask method to match
keywords.

sarrrdin committed:
#Add extension C-Priority

Users want to be able to set priority of their tasks, so they can
see what needs to be done first.

Priority command function is added to Duke, so users can simply
command:
```
priority <priority level> /for <index of task in list to update>
```
(default priority level is 0)

As number of command increases, HiCommand is added so users can
see the list of instructions. Users can simply type
```
hi
```
to see the functionality of Duke.

sarrrdin committed:
Update some exception handling, such as index out of bounds. Check
code and tidy up + optimization to help improve readability and
sustainability.

sarrrdin committed:
Set up CI

sarrrdin committed:
Improve GUI, change chat colour, font and layout

sarrrdin committed:
Change personality of Duke (different from original instructions). Now
is called DukePro and not Duke!

sarrrdin committed:
Fix bugs for DeleteCommand, and certain formatting issues for better Ui

sarrrdin committed:
Add README file and Ui for DukePro

sarrrdin committed:
Merge pull request #2 from sarrrdin/branch-Published

Add README file and Ui for DukePro

sarrrdin committed:
Fine tune byeCommand such that program exits

sarrrdin committed:
Tweak abit on README to make it clearer.

sarrrdin committed:
Revert "Tweak abit on README to make it clearer."

This reverts commit 617ef559ed2e6cb1fd9cbd0bf468418c2cdb6aee.

sarrrdin committed:
Tweak README.md to make it more comprehensive

sarrrdin committed:
Change welcome message to DukePro
